File: py/portfolio.py
import functools
import operator

from money import Money


class Portfolio:
    def __init__(self, *moneys):
        self.moneys = []
        self._eur_to_usd = 1.2
        self.moneys.extend(moneys)

    def __convert(self, aMoney, aCurrency):
        if aMoney.currency == aCurrency:
            return aMoney.amount
        return aMoney.amount * self._eur_to_usd

    def evaluate(self, currency):
        return Money(
            functools.reduce(
                operator.add, map(lambda m: self.__convert(m, currency), self.moneys), 0
            ),
            currency,
        )

    def add(self, *moneys):
        self.moneys.extend(moneys)

File: js/portfolio.js
import Money from "./money.js";

function createErrorMessage(failures) {
  return failures.map(failure => failure.reason).join();
}

export default class Portfolio {
  #money;

  constructor(...money) {
    this.#money = money || [];
  }

  add(...money) {
    this.#money = this.#money.concat(money);
    return this;
  }

  async evaluate(aCurrency, exchange) {
    const converted = await Promise.allSettled(
      this.#money.map(aMoney => exchange.convert(aMoney, aCurrency))
    );
    const failures = converted.filter(result => result.status === "rejected");
    if (failures.length === 0) {
      return converted.reduce(
        (acc, result) => acc.add(result.value),
        new Money(0, aCurrency)
      );
    }
    throw new Error(
      `Missing exchange rate(s): [${createErrorMessage(failures)}]`
    );
  }
}
